src/semantic/types/constraints/PolyTypeConstraintVisitor.cpp

```diff
@@ -0,0 +1,64 @@
+#include "PolyTypeConstraintVisitor.h"
+#include "FreshAlphaCopier.h"
+#include "TypeVars.h"
+#include "loguru.hpp"
+
+PolyTypeConstraintVisitor::PolyTypeConstraintVisitor(
+    SymbolTable *st, CallGraph* cg, std::shared_ptr<Unifier> u, 
+    std::unique_ptr<ConstraintHandler> handler)
+    : TypeConstraintVisitor(st, std::move(handler)), callGraph(cg), unifier(u) {};
+
+/*! \brief Polymorphic type constraints for function application.
+ *
+ * Type Rules for "E(E1, ..., En)":
+ *  [[E]] = ([[E1]], ..., [[En]]) -> [[E(E1, ..., En)]]
+ *
+ * Polymorphic typing handles the case where expression "E" has
+ * a function type involving free type variables, i.e., \alpha.
+ * In this case, rather than force such type variables to be unified
+ * across all call sites, instead we create a copy of the function
+ * type with unique free type variables, i.e., the instantiated type,
+ * that can be unified with the function expression in the function application.
+ *
+ * An additional complexity is that we will perform this process for each
+ * function that may be called from this call site.  Since unification solves
+ * the conjunction of constraints this forces all of the corresponding free
+ * type variables for each called function to be consistently typed at this call
+ * site.  Different typings are possible at other call sites.
+ *
+ * FIX THE ABOVE ONCE THINGS ARE WORKING
```

Comment:
Should this line be removed?

Reply:
Nice catch.  This required a substantial rewrite and I've done that now.